File: src/Samurai.Client.Console/Samurai.Client.Console/ServerApi.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Text;
using Newtonsoft.Json;
using SamuraiServer.Data;

namespace Samurai.Client.ConsoleClient
{
    public class ServerApi
    {
        private readonly string serverUrl;

        public ServerApi(string serverUrl)
        {
            this.serverUrl = serverUrl;
        }

        public void CreatePlayer(string name, Action<CreatePlayerResponse, Exception> callback)
        {
            Post("/Api/Players/CreatePlayer", "name=" + System.Uri.EscapeDataString(name), callback);
        }

        public void Login(string name, string key, Action<CreatePlayerResponse, Exception> callback)
        {
            Post("/Api/Players/Login", "name=" + System.Uri.EscapeDataString(name) + "&token=" + System.Uri.EscapeDataString(key), callback);
        }

        public void GetOpenGames(Action<GetOpenGamesResponse, Exception> callback)
        {
            Get("/Api/Games/GetOpenGames", callback);
        }

        public void CreateGameAndJoin(string name, Guid playerId, Action<CreateGameAndJoinResponse, Exception> callback)
        {
            Post("/Api/Games/CreateGameAndJoin", "name=" + System.Uri.EscapeDataString(name) + "&playerid=" + playerId, callback);
        }

        public void JoinGame(Guid gameId, Guid playerId, Action<CreateGameAndJoinResponse, Exception> callback)
        {
            Post("/Api/Games/JoinGame", "gameid=" + gameId + "&playerid=" + playerId, callback);
        }

        public void GetMap(Guid mapId, Action<GetMapResponse, Exception> callback)
        {
            Post("/Api/Games/GetMap", "mapid=" + mapId, callback);
        }

        public void StartGame(Guid gameId, Action<StartGameResponse, Exception> callback) {
            Get("/Api/Games/StartGame?gameid=" + gameId, callback);
        }

        // A sync method in an async pattern. Easier to work with for the console app.
        private void Get<T>(string url, Action<T, Exception> callback)
        {
            try
            {
                var request = (HttpWebRequest)HttpWebRequest.Create(Uri(url));
                request.Accept = "application/json";
                var response = (HttpWebResponse)request.GetResponse();
                if (response.StatusCode != HttpStatusCode.OK) callback(default(T), new Exception("Unexpected server response " + response.StatusCode));

                string json;
                using (var reader = new StreamReader(response.GetResponseStream()))
                {
                    json = reader.ReadToEnd();
                    reader.BaseStream.Close();
                }
                response.Close();

                var data = JsonConvert.DeserializeObject<T>(json);
                callback(data, null);
            }
            catch (Exception e)
            {
                callback(default(T), e);
            }
        }

        // A sync method in an async pattern. Easier to work with for the console app.
        private void Post<T>(string url, string data, Action<T, Exception> callback)
        {
            try
            {
                byte[] bytes = Encoding.UTF8.GetBytes(data);
                var request = (HttpWebRequest)HttpWebRequest.Create(Uri(url));
                request.Method = "POST";
                request.Accept = "application/json";
                request.ContentType = "application/x-www-form-urlencoded";
                using (var stream = request.GetRequestStream())
                {
                    stream.Write(bytes, 0, bytes.Length);
                }
                var response = (HttpWebResponse)request.GetResponse();
                if (response.StatusCode != HttpStatusCode.OK) callback(default(T), new Exception("Unexpected server response " + response.StatusCode));

                string json;
                using (var reader = new StreamReader(response.GetResponseStream()))
                {
                    json = reader.ReadToEnd();
                    reader.BaseStream.Close();
                }
                response.Close();

                var responseData = JsonConvert.DeserializeObject<T>(json);
                Debug.WriteLine(String.Format("Request: {0} at {1}", url, DateTime.Now));
                Debug.WriteLine(String.Format("Data: {0}", data));
                Debug.WriteLine("Response:");
                Debug.WriteLine(json);
                Debug.WriteLine("=========================================================");
                callback(responseData, null);
            }
            catch (Exception e)
            {
                callback(default(T), e);
            }
        }

        private Uri Uri(string relativePath)
        {
            return new Uri(new Uri(serverUrl), relativePath);
        }
    }

    public class ServerResponse
    {
        public bool Ok { get; set; }
        public string Message { get; set; }
    }

    public class GetOpenGamesResponse : ServerResponse
    {
        public GameState[] Games { get; set; }
    }

    public class CreateGameAndJoinResponse : ServerResponse
    {
        public GameState Game { get; set; }
    }

    public class JoinGameResponse : ServerResponse
    {
        public GameState Game { get; set; }
    }

    public class CreatePlayerResponse : ServerResponse
    {
        public Player Player { get; set; }
    }

    public class LoginResponse : ServerResponse
    {
        public Player Player { get; set; }
    }

    public class GetMapResponse : ServerResponse
    {
        public string[] Map { get; set; }
    }

    public class StartGameResponse : ServerResponse
    {
        public GameState Game { get; set; }
    }
}

File: src/Samurai.Client.JS/serverApi.js
define(['gameState', 'player'], function (GameState, Player) {
  /** Server API module.
   * Ported from C# implementation.
   */
  function ServerApi(serverUrl) {
    var serverUrl = serverUrl;


    function createPlayer(name, callback) {
      post("api/players/createPlayer", {name : name}, callback);
    }


    function login(name, key, callback) {
      post("api/players/login", {name : name, token : key}, callback);
    }


    function getOpenGames(callback) {
      get("api/games/getOpenGames", callback);
    }


    function createGameAndJoin(name, playerId, callback) {
      post("api/games/createGameAndJoin", {name : name, playerid : playerId}, callback);
    }


    function joinGame(gameId, playerId, callback) {
      post("api/games/joinGame", {gameid : gameId, playerid : playerId}, callback);
    }


    function getMap(mapId, callback) {
      post("api/games/getMap", {mapid : mapId}, callback);
    }

    function startGame(gameId, callback) {
      get("api/games/startGame", callback, {gameid: gameId});
    }

    function getGame(gameId, callback) {
      get("api/games/getGame", callback, {gameid: gameId});
    }

    //Synchronous GET request.
    function get(url, callback, data) {
      try {
        $.ajax(serverUrl + url, {
          async: false,
  //        contentType: "application/json",
          dataType: "json",
          data: data,
          success: function (data) {
            callback(data, null);
          },
          error: function (jqXHR, textStatus) {
  //          if (response.StatusCode != HttpStatusCode.OK) callback(default(T), new Exception("Unexpected server response " + response.StatusCode));
            callback(null,  'Unexpected server response ' + jqXHR.status);
          },
        });
      } catch(e) {
        console.log('Caught an error in serverApi get: ' + e.message);
        callback(null, e);
      }
    }

    function post(url, data, callback) {
      try {
        $.ajax(serverUrl + url, {
          type: "POST",
          async: false,
  //        contentType: "application/json",
          dataType: "json",
          data: data,
          success: function (responseData) {
              callback(responseData, null);
           },
          error: function (jqXHR, textStatus) {
            console.log('Request: ' + url + ' at ' + Date.now());
            console.log('Data: ', data);
            console.log('Textstatus', textStatus);
            callback(null, 'Unexpected server response ' + jqXHR.status);
          }
        });
      } catch(e) {
        console.log('Caught an error in serverApi post: ' + e.message);
        callback(null, e);
      }
    }

    return {
       createPlayer : createPlayer,
       login : login,
       getOpenGames : getOpenGames,
       createGameAndJoin : createGameAndJoin,
       joinGame : joinGame,
       getMap : getMap,
       startGame : startGame,
       getGame : getGame,
    };
  }

  /*****************************************************************************/
  function ServerResponse() {
  //    public bool Ok { get; set; }
  //    public string Message { get; set; }
    this.ok = false;
    this.message = '';
  }

  function GetOpenGamesResponse() {
  //    public GameState[] Games { get; set; }
    ServerResponse.call(this);
    this.Games = [];
  }
  GetOpenGamesResponse.prototype = new ServerResponse();
  GetOpenGamesResponse.prototype.constructor = GetOpenGamesResponse();

  function CreateGameAndJoinResponse() {
  //    public GameState Game { get; set; }
    ServerResponse.call(this);
    this.game = new GameState();
  }
  CreateGameAndJoinResponse.prototype = new ServerResponse();
  CreateGameAndJoinResponse.prototype.constructor = CreateGameAndJoinResponse();

  function JoinGameResponse() {
  //    public GameState Game { get; set; }
    ServerResponse.call(this);
    this.game = new GameState();
  }
  JoinGameResponse.prototype = new ServerResponse();
  JoinGameResponse.prototype.constructor = JoinGameResponse();

  function CreatePlayerResponse() {
  //    public Player Player { get; set; }
    ServerResponse.call(this);
    this.Player = new Player();
  }
  CreatePlayerResponse.prototype = new ServerResponse();
  CreatePlayerResponse.prototype.constructor = CreatePlayerResponse();

  function LoginResponse() {
  //    public Player Player { get; set; }
    ServerResponse.call(this);
  //  this.Player = new Player();
    this.Player = null;
  }
  LoginResponse.prototype = new ServerResponse();
  LoginResponse.prototype.constructor = LoginResponse();

  function GetMapResponse() {
  //    public string[] Map { get; set; }
    ServerResponse.call(this);
    this.map = [];
  }
  GetMapResponse.prototype = new ServerResponse();
  GetMapResponse.prototype.constructor = GetMapResponse();

  function StartGameResponse() {
  //    public GameState Game { get; set; }
    ServerResponse.call(this);
    this.game = new GameState();
  }
  StartGameResponse.prototype = new ServerResponse();
  StartGameResponse.prototype.constructor = StartGameResponse();

  function GetGameResponse() {
  //    public GameState Game { get; set; }
    ServerResponse.call(this);
    this.game = new GameState();
  }
  GetGameResponse.prototype = new ServerResponse();
  GetGameResponse.prototype.constructor = GetGameResponse();
  /*****************************************************************************/

  return ServerApi;
});